Pass qtile manager to all hooks

m-col opened the pull request.

m-col commented:
Changes thus far:

- passing `qtile` every time a hook is fired, as the first argument, automatically
- updating `hook.fire` instances which manually passed `qtile`
- updating hooked functions' arguments

We could also:

- tidy any hooked functions within libqtile that use `qtile` with a nested reference, where we could use the passed `qtile` directly
- remove the deprecated `window_name_change` hook

tych0 commented:
I thought per https://github.com/qtile/qtile/issues/1719#issuecomment-647624197 that we were going to move libqtile.hook.qtile to somewhere more sensible instead?

m-col commented:
My bad, I must have missed that last comment. If we moved that qtile global somewhere else, would we then want to access it within libqtile.hooks to pass it to hooked functions, or actually just remove the qtile instance from all hook arguments and let the hooked functions deal with getting the qtile instance themselves (if they need it)?

tych0 commented:
On Mon, Jun 29, 2020 at 09:41:01AM -0700, m-col wrote:
> My bad, I must have missed that last comment. If we moved that qtile global somewhere else, would we then want to access it within libqtile.hooks to pass it to hooked functions, or actually just remove the qtile instance from all hook arguments and let the hooked functions deal with getting the qtile instance themselves (if they need it)?

It seems nicer to just have it be global? But globals are also
terrible. I could be convinced either way :)

m-col commented:
I think this usage of global is pretty restricted anyway, so it seems like an easy way to get an importable reference to `qtile`. I'd argue the neatest place would be at `libqtile.qtile` by moving the `hooks.init` function into libqtile's `__init__.py`. This might be a nice way to get the first suggestion in #1751 working. I think that #1751 might have to be completed before updating the hooks, so that the hooks can use that new API.

tych0 commented:
> I think that #1751 might have to be completed before updating the hooks, so that the hooks can use that new API.

I don't think #1751 is coming any time soon; I don't even really understand how to implement it :). Let's not block any hooks fixups on that.

If you think a global is better, do you want to switch this PR to a global (and maybe add a doc somewhere? :D)

m-col commented:
Actually yeah - if/when #1751 is worked on, the API would presumably be
unchanged i.e. `import qtile from libqtile`.

So before I make some updates, would this be the plan for this PR:

  - move hook.init and hook.qtile so qtile is available at libqtile.qtile
  - update any hook.fire calls to remove qtile from the arguments
  - import qtile to any modules with hooked functions that currently get qtile
    from hook.fire arguments

Perhaps we should also add a deprecation warning to the use of `main` in
configs?

On 200630 09:30, Tycho Andersen wrote:
>> I think that #1751 might have to be completed before updating the hooks, so that the hooks can use that new API.
>
>I don't think #1751 is coming any time soon; I don't even really understand how to implement it :). Let's not block any hooks fixups on that.
>
>If you think a global is better, do you want to switch this PR to a global (and maybe add a doc somewhere? :D)
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/qtile/qtile/pull/1791#issuecomment-651905781

ramnes commented:
Yep, I'm in line with everything in these last two comments. :)

tych0 commented:
> Perhaps we should also add a deprecation warning to the use of `main` in configs?

Yes, and a bit of documentation on how to access the qtile object on the hooks page would be helpful too. Thanks!

m-col commented:
Test failed with "libqtile/hook.py:38:1: I004 isort found an unexpected blank line in imports"

Is it meant to be 1 or 2 blank lines after imports? I thought it was 2.

ramnes commented:
It's 2 blank lines around functions and classes, and 1 otherwise.

https://www.python.org/dev/peps/pep-0008/#blank-lines

m-col commented:
Huh, TIL, I am ashamed.

m-col commented:
These are the changes:

- moved hook.init and hook.qtile so qtile is available at libqtile.qtile
- updated hook.fire calls to remove qtile from the arguments
- import qtile to any modules with hooked functions that currently get qtile from hook.fire arguments
- add a deprecation warning to the use of main in configs
- added documentation on how to access the qtile object on the hooks page
- removed `main` from the `configuration variables` docs

Using a main function will still work but you'll get a deprecation warning in the log. Anything done in `main` can just be done directly in the config with `libqtile.qtile`.

Config breakage: hooked functions will not get passed the `qtile` object if they previously did.

ramnes commented:
Can you add this to the CHANGELOG?

m-col commented:
Done. I also noticed `main` in the test configs and removed those. However, there was one thing in the tests that I was unsure how to update. [Here](https://github.com/qtile/qtile/blob/master/test/test_manager.py#L1134), [here](https://github.com/qtile/qtile/blob/master/test/test_manager.py#L1160) and [here](https://github.com/qtile/qtile/blob/master/test/layouts/test_common.py#86) there are main functions defined in these test configs, though I'm not familiar enough with pytest to know exactly how to have the equivalent functionality outside of a main function. Any thoughts?

tych0 commented:
Can you remove the commits that cancel each other out (adding qtile to hooks, then removing it) and rebase? The other stuff lgtm.

ramnes commented on libqtile/core/manager.py:
```diff
@@ -108,6 +109,7 @@ def __init__(
 
         self.core.wmname = getattr(self.config, "wmname", "qtile")
         if config.main:
+            warnings.warn("Defining a main function is deprecated, use libqtile.qtile", DeprecationWarning)
```
Should we mention `@hook.subscribe.startup_complete` also?

ramnes replied:
(here and in the changelog)

m-col replied:
To remind people that `qtile` can also be modified after startup?

ramnes replied:
If you try to play `libqtile.qtile` inside the config directly, IIRC most things won't be initiated yet, so you will very quickly end up with errors. But if you play with it inside `@hook.subscribe.startup_complete`, you have the guarantee that everything is instantiated.

ramnes commented:
Or more simply, let's squash everything down to one commit. :)

m-col commented:
How is the updated changelog entry?

Tests passed and rebased.

ramnes commented on CHANGELOG:
```diff
@@ -1,3 +1,13 @@
+	- hooks that pass the qtile object as an argument no longer will. It
+	  can be accessed directly at libqtile.qtile.
+    !!! deprecation warning !!!
+	- defining a main function in your config is deprecated. Instead the
+	  qtile object can be accessed before setup by importing libqtile.qtile
+	  into your config, or after setup by using
+	  hook.subscribe.startup_complete.
```
```suggestion
	- defining a main function in your config is deprecated. You should use
	  @hook.subscribe.startup_complete instead. If you need access to the
	  qtile object, import it from libqtile directly.
```

ramnes replied:
Also, it's a config breakage rather than a deprecation now that you removed it from configreader.py, isn't it?

m-col replied:
I jumped the gun a bit with that one, and misunderstood how the confreader works. You're right. The intended behaviour was:
```
if the config defines main:
    save it as qtile.main, and run it upon start with the deprecation warning
else:
    do nothing with main
````
We don't want `main` in the default config if it's being deprecated, but still want it to be allowed. Maybe we could just add a comment for now in the default config saying it will soon be deprecated, do not use?

tych0 commented:
I still see a whole bunch of commits that cancel each other out. Can you drop those?

m-col commented:
OK hopefully that is it :)

tych0 commented on CHANGELOG:
```diff
@@ -1,3 +1,8 @@
+qtile X.X.X, released XXXX-XX-XX:
+    !!! Config breakage !!!
+	- hooks that pass the qtile object as an argument no longer will. It
```
Can you add the list of the hooks here so people can grep?

m-col replied:
Ah yep good idea
